state-chain/pallets/cf-swapping/src/tests/dca.rs

```diff
@@ -0,0 +1,582 @@
+
+const INPUT_AMOUNT: AssetAmount = 40_000;
+const CHUNK_INTERVAL: u32 = 3;
+const BROKER_FEE: AssetAmount = INPUT_AMOUNT * BROKER_FEE_BPS as u128 / 10_000;
+const NET_AMOUNT: AssetAmount = INPUT_AMOUNT - BROKER_FEE;
+
+const INPUT_ASSET: Asset = Asset::Usdc;
+const OUTPUT_ASSET: Asset = Asset::Eth;
+
+fn params(
+	dca_params: Option<DcaParameters>,
+	refund_params: Option<SwapRefundParameters>,
+	is_ccm: bool,
+) -> TestSwapParams {
+	TestSwapParams {
+		input_asset: INPUT_ASSET,
+		output_asset: OUTPUT_ASSET,
+		input_amount: INPUT_AMOUNT,
+		refund_params,
+		dca_params,
+		output_address: (*EVM_OUTPUT_ADDRESS).clone(),
+		is_ccm,
+	}
+}
+
+#[test]
+fn dca_happy_path() {
+	const CHUNK_1_BLOCK: u64 = 3;
+	const CHUNK_2_BLOCK: u64 = CHUNK_1_BLOCK + CHUNK_INTERVAL as u64;
+
+	const CHUNK_AMOUNT: AssetAmount = NET_AMOUNT / 2;
+
+	// 1:1 swap ratio
+	const TOTAL_OUTPUT_AMOUNT: AssetAmount = NET_AMOUNT;
+
+	new_test_ext()
+		.execute_with(|| {
+			assert_eq!(System::block_number(), INIT_BLOCK);
+
+			insert_swaps(&[params(
+				Some(DcaParameters { number_of_chunks: 2, chunk_interval: CHUNK_INTERVAL }),
+				None,
+				false,
+			)]);
+
+			assert_has_matching_event!(
+				Test,
+				RuntimeEvent::Swapping(Event::SwapRequested {
+					swap_request_id: SWAP_REQUEST_ID,
+					input_amount: INPUT_AMOUNT,
+					..
+				})
+			);
+
+			assert_has_matching_event!(
+				Test,
+				RuntimeEvent::Swapping(Event::SwapScheduled {
+					swap_request_id: SWAP_REQUEST_ID,
+					swap_id: 1,
+					input_amount: CHUNK_AMOUNT,
+					execute_at: CHUNK_1_BLOCK,
+					..
+				})
+			);
+		})
+		.then_execute_at_block(CHUNK_1_BLOCK, |_| {})
+		.then_execute_with(|_| {
+			assert_has_matching_event!(
+				Test,
+				RuntimeEvent::Swapping(Event::SwapExecuted {
+					swap_request_id: SWAP_REQUEST_ID,
+					swap_id: 1,
+					input_amount: CHUNK_AMOUNT,
+					output_amount: CHUNK_AMOUNT,
+					..
+				})
+			);
+
+			// Second chunk should be scheduled 2 blocks after the first is executed:
+			assert_has_matching_event!(
+				Test,
+				RuntimeEvent::Swapping(Event::SwapScheduled {
+					swap_request_id: SWAP_REQUEST_ID,
+					swap_id: 2,
+					input_amount: CHUNK_AMOUNT,
+					execute_at: CHUNK_2_BLOCK,
+					..
+				})
+			);
+		})
+		.then_execute_at_block(CHUNK_2_BLOCK, |_| {})
+		.then_execute_with(|_| {
+			assert_eq!(SwapRequests::<Test>::get(SWAP_REQUEST_ID), None);
+
+			assert_event_sequence!(
+				Test,
+				RuntimeEvent::Swapping(Event::SwapExecuted {
+					swap_request_id: SWAP_REQUEST_ID,
+					swap_id: 2,
+					input_amount: CHUNK_AMOUNT,
+					output_amount: CHUNK_AMOUNT,
+					..
+				}),
+				RuntimeEvent::Swapping(Event::SwapEgressScheduled {
+					swap_request_id: SWAP_REQUEST_ID,
+					amount: TOTAL_OUTPUT_AMOUNT,
+					..
+				}),
+				RuntimeEvent::Swapping(Event::SwapRequestCompleted {
+					swap_request_id: SWAP_REQUEST_ID
+				}),
+			);
+		});
+}
+
+/// Test that DCA with 1 chunk behaves like a regular swap
+#[test]
+fn dca_single_chunk() {
```

Comment:
Sure. Still, I feel like it is nicer to have an explicit test? Otherwise, it is not immediately clear that this case is (implicitly) covered by non-DCA tests. Can remove if you feel strongly about it though.

Reply:
Fine to leave it - as mentioned above, I think the tests are due a refactor anyway.